File: backend/app/services/AnalyticsService.py
from datetime import datetime
from sqlalchemy.orm import Session

from app.models.RoleModel import RoleEnum as RE
from app.models.TimeLogModel import TypeEnum
from app.repositories.AnalyticsRepository import AnalyticsRepository
import app.schemas.AnalyticsSchema as AnalyticsSchema
from app.core.exceptions import AdminAccessRequiredError, ManagerAccessRequiredError


class AnalyticsService:

    @staticmethod
    def get_reports(
        view_as: list[ RE ],
        start_date: datetime,
        end_date: datetime,
        project_ids: list[int] | None,
        type_filters: list[TypeEnum],
        sort_by: str,
        sort_type: int,
        current_user: dict,
        db: Session,
    ) -> list[AnalyticsSchema.ReportResponse]:
        user_roles = current_user.get("user_roles", [])

        rows = AnalyticsRepository.get_report_data(
            view_as=view_as,
            start_date=start_date,
            end_date=end_date,
            project_ids=project_ids,
            type_filters=type_filters,
            sort_by=sort_by,
            sort_type=sort_type,
            current_user_id=current_user["user_id"],
            db=db,
        )

        return [
            AnalyticsSchema.ReportResponse(
                id=row.id,
                name=row.name or "",
                userid=row.userid or "",
                project_id=row.project_id,
                project_name=row.project_name or "",
                hours=round(float(row.total_minutes or 0) / 60.0, 2),
            )
            for row in rows
        ]

File: frontend/src/services/analyticsService.js
import { analyticsApi } from '../apis/analyticsApi';

export const analyticsService = {
  getReports: async (filters = {}) => {
    const params = {
      sort_by: filters.sortBy || 'duration',
      sort_type: filters.sortType || -1,
    };
    if (filters.viewAs && filters.viewAs.length > 0) {
      params.view_as = filters.viewAs;
    }
    if (filters.startDate) params.start_date = filters.startDate;
    if (filters.endDate) params.end_date = filters.endDate;
    if (filters.projectIds && filters.projectIds.length > 0) {
      params.project_ids = filters.projectIds;
    }
    if (filters.type && filters.type.length > 0) {
      params.type = filters.type;
    }
    return await analyticsApi.getReports(params);
  },
};
